android-core/src/main/java/com/mparticle/internal/DeviceAttributes.kt

```diff
@@ -0,0 +1,343 @@
+                    attributes.put(MessageKey.MOBILE_COUNTRY_CODE, networkOperator.substring(0, 3))
+                    attributes.put(MessageKey.MOBILE_NETWORK_CODE, networkOperator.substring(3))
+                }
+            }
+            attributes.put(MessageKey.DEVICE_IS_TABLET, MPUtility.isTablet(appContext))
+            attributes.put(MessageKey.DEVICE_IS_IN_DST, MPUtility.isInDaylightSavings())
+
+            if (!MPUtility.isEmpty(deviceImei)) {
+                attributes.put(MessageKey.DEVICE_IMEI, deviceImei)
+            }
+        } catch (e: Exception) {
+            // believe it or not, difference devices can be missing build.prop fields, or have otherwise
+            // strange version/builds of Android that cause unpredictable behavior
+        }
+
+        return attributes
+    }
+
+    /**
+     * For the following fields we always want the latest values
+     */
+    fun updateDeviceInfo(context: Context, deviceInfo: JSONObject) {
+        deviceInfo.remove(MessageKey.LIMIT_AD_TRACKING)
+        deviceInfo.remove(MessageKey.GOOGLE_ADV_ID)
+        val adIdInfo = MPUtility.getAdIdInfo(context)
+        var message =
+            "Failed to collect Advertising ID, be sure to add Google Play services (com.google.android.gms:play-services-ads) or Amazon Ads (com.amazon.android:mobile-ads) to your app's dependencies."
+        if (adIdInfo != null) {
+            try {
+                deviceInfo.put(MessageKey.LIMIT_AD_TRACKING, adIdInfo.isLimitAdTrackingEnabled)
+                val instance = MParticle.getInstance()
+                // check instance nullability here and decline to act if it is not available. Don't want to have the case where we are overriding isLimiAdTrackingEnabled
+                // just because there was a timing issue with the singleton
+                if (instance != null) {
+                    if (adIdInfo.isLimitAdTrackingEnabled) {
+                        message = adIdInfo.advertiser.descriptiveName + " Advertising ID tracking is disabled on this device."
+                    } else {
+                        when (adIdInfo.advertiser) {
+                            MPUtility.AdIdInfo.Advertiser.AMAZON -> deviceInfo.put(MessageKey.AMAZON_ADV_ID, adIdInfo.id)
+                            MPUtility.AdIdInfo.Advertiser.GOOGLE -> deviceInfo.put(MessageKey.GOOGLE_ADV_ID, adIdInfo.id)
+                        }
+                        message = "Successfully collected " + adIdInfo.advertiser.descriptiveName + " Advertising ID."
+                    }
+                }
+            } catch (jse: JSONException) {
+                Logger.debug("Failed while building device-customAttributes object: ", jse.toString())
+            }
+        }
+        if (firstCollection) {
+            Logger.debug(message)
+            firstCollection = false
+        }
+
+        try {
+            val mParticle = MParticle.getInstance()
+            if (mParticle != null) {
```

Comment:
Same here

Reply:
We can't use let here because, up until line 275, we'd have to check for null on every line if we did.